source/agora/consensus/Fee.d

```diff
@@ -353,85 +346,105 @@ public class FeeManager
     {
-        Amount tot_fee, tot_data_fee;
-        this.getTXSetFees(block.txs, peekUTXO, tot_fee, tot_data_fee);
+        if (block.header.height == 0) // No fees for Genesis Block
+            return;
 
-        // Filter out the missing validators as they will not get paid fees for this block
-        auto vals_to_pay = validators.enumerate.filter!(en =>
-            !block.header.missing_validators.canFind(en.index)).map!(en => en.value).array;
+        Amount txs_fees;
+        Amount txs_data_fees;
+        this.getTXSetFees(block.txs, peekUTXO, txs_fees, txs_data_fees);
 
-        const validator_fees = this.getValidatorFees(tot_fee, tot_data_fee, vals_to_pay);
+        this.block_fees[block.header.height] = txs_fees;
+        this.block_data_fees[block.header.height] = txs_data_fees;
+        this.db.execute(
+                "REPLACE INTO block_fees (height, fee, data_fee) VALUES (?, ?, ?)",
+                block.header.height, txs_fees, txs_data_fees);
+    }
 
-        const commons_fees = this.getCommonsBudgetFee(tot_fee, tot_data_fee, vals_to_pay);
+    /***************************************************************************
 
-        foreach (idx, validator; vals_to_pay)
-        {
-            this.accumulated_fees.update(validator.address,
-                { return validator_fees[idx]; },
-                (ref Amount so_far) {
-                    so_far += validator_fees[idx];
-                    return so_far;
-                }
-            );
-
-            this.db.execute(
-                "REPLACE INTO accumulated_fees (fee, public_key) VALUES (?,?)",
-                this.accumulated_fees[validator.address].tupleof[0], validator.address);
-        }
-        this.accumulated_fees.update(this.params.CommonsBudgetAddress,
-                { return commons_fees; },
-                (ref Amount so_far) {
-                    so_far += commons_fees;
-                    return so_far;
-                }
-            );
-        this.db.execute(
-                "REPLACE INTO accumulated_fees (fee, public_key) VALUES (?,?)",
-                this.accumulated_fees[this.params.CommonsBudgetAddress].tupleof[0], this.params.CommonsBudgetAddress);
+        Returns the total `Amount` of fees in the transaction set of a block
+
+        This includes any `data_fees` which are based on the size of the
+        `payload` in a `Transaction`.
+        The transaction fee rate is calculated after the data fee is subtracted
+        from the total fees, which is `sum of transactions input values`
+        minus `sum of output values`.
+
+        Params:
+            height = height of block
+
+        Return:
+            `Amount` of fees included at given height
+
+    ***************************************************************************/
+
+    public Amount getBlockFees (in Height height) nothrow @safe
+    {
+        if (height !in this.block_fees)
+            return 0.coins;
+        return this.block_fees[height];
```

Comment:
This is doing a double lookup.
```suggestion
        if (auto ptr = height in this.block_fees)
            return *ptr;
        return 0.coins;
```

Or simply:
```suggestion
        return this.block_fees.get(height, 0.coins);
```

Reply:
I like the second one.